contracts/rosca.sol

```diff
@@ -1,266 +1,258 @@
+   *
+   */
+  function joinRequest() beforeStart
+  {
+    //only put the request in the pending list if they are not in the ROSCA already
+    if(!members[msg.sender].alive)
+      pendingJoinRequest[msg.sender] = true;
+    else
+      throw;
+  }
+  /**
+   * foreman must pass in the address of the requtor to approve the ROSCA participation
+   * once a requestor had been registered as member, the address will be taken out of pendingJoinRequest
+   **/
+  function acceptJoinRequest(address requestor)
+    onlyForeman
+    beforeStart
+  {
+    if(pendingJoinRequest[requestor])
     {
-        if(distrubtionAmountInWei < lowestBid || !(members[msg.sender].paid < currentEpoch))
-        {
-            //set new lowestBid and winnerAddress, also trigger event new Bid placed
-            lowestBid = distrubtionAmountInWei;
-            winnerAddress = msg.sender;
-            LogNewLowestBid(lowestBid, winnerAddress);
-
-        }else
-            throw;
+      members[requestor] = USER({paid: 0 , contributed: 0, alive: true, pendingWithdrawl: 0});
+      membersAddresses.push(requestor);
+      numRoundsInEpoch++;
+      LogParticipantRegistered(requestor);
+      delete(pendingJoinRequest[requestor]); // take out the requestor's address in the pending list
     }
-
-    /**
-     * Withdraws available funds for msg.sender. If opt_destination is specified,
-     * sends the fund to that address.
-     */
-    function withdraw(address opt_destination) returns(bool success){
-
-        if(members[msg.sender].alive && members[msg.sender].allowance > 0)
-        {
-            uint amountToWithdraw = members[msg.sender].allowance;  //use a temporary variable to avoid the receiver calling the withdraw function again before
-            members[msg.sender].allowance = 0;                      // send() completes its task
-
-
-            if (!opt_destination.send(amountToWithdraw)) {   //if the send() fails, put the allowance back to its original place
-                // No need to call throw here, just reset the amount owing
-                members[msg.sender].allowance = amountToWithdraw;
-                return false;
-            //TODO(shine) : look into security about withdraw
-            }
-            else{
-                LogFundsWithdrawal(msg.sender, amountToWithdraw, opt_destination);
-                return true;
-            }
-        }
-        else
-            throw;
-
-    }
-
-    function integrityCheck(uint option) // option will specify whether it will be a message call or transaction call to freeze the contract
+    else throw;
+  }
+
+  // v2: function leave();
+
+  /**
+   * Processes a periodic contribution from msg.sender ().
+   * Any excess funds will be withdrawable through withdraw().
+   */
+  function contribute() payable
+  {
+    //Product condsideration : are we gonna limit how much ether could be sent? is it total, or how much sent each time?
```

Comment:
We can throw if a user sends too much, but I prefer we just credit them (e.g. put it in pendingWithdrawls) - this will allow users to e.g. contribute 2 rounds' worth of money at one time, if they prefer.